python_base/docker_entrypoint.sh

```diff
@@ -21,19 +21,76 @@
 # In applying this licence, CERN does not waive the privileges and immunities
 # granted to it by virtue of its status as an Intergovernmental Organization
 # or submit itself to any jurisdiction.
+#
+# Note on the usage of signal handlers.
+#
+# In order to be able to restore the ownership of the venv at the end, we have
+# to declare it as a trap for the EXIT signal, that forces us to not use
+# 'exec' for the wrapped command, and that leads us to manually having to
+# forward also the SIGTERM and SIGINT signals.
+#
+# Also, to have job control (to be able to run 'fg') bash needs to be set in
+# monitor mode (-m)
+set -me
+
+VENV_PATH=/virtualenv
+
+
+restore_venv_rights() {
+    if [[ "$BASE_USER_UID" != "" ]]; then
+        BASE_USER_GID="${BASE_USER_GID:-$BASE_USER_UID}"
+        echo "Restoring permissions or venv to $BASE_USER_UID:$BASE_USER_GID"
+        /fix_venv_rights "$BASE_USER_UID:$BASE_USER_GID"
+    else
+        echo "No BASE_USER_UID env var defined, skipping venv permission" \
+            "restore."
+    fi
+}
+
+forward_sigterm() {
+    echo "Forwarding SIGTERM to $child"
+    kill -SIGTERM "$child" &>/dev/null
+    trap forward_sigterm SIGTERM
+    wait "$child"
+}
 
-set -e
 
-if [ ! -f /virtualenv/bin/activate ]; then
-    virtualenv /virtualenv -p python${INSPIRE_PYTHON_VERSION}
-    source /virtualenv/bin/activate
+forward_sigint() {
+    echo "Forwarding SIGINT to $child"
+    kill -SIGINT "$child" &>/dev/null
+    trap forward_sigterm SIGINT
+    wait "$child"
+}
+
+
+prepare_venv() {
+    virtualenv "$VENV_PATH" -p "python${INSPIRE_PYTHON_VERSION}"
+    source "$VENV_PATH"/bin/activate
     pip install --upgrade pip
     pip install --upgrade setuptools wheel
-    cp -r /src-cache /virtualenv/src
-else
-    source /virtualenv/bin/activate
-fi
+    cp -r /src-cache "$VENV_PATH"/src
+}
+
+
+main() {
+    /fix_venv_rights 'test:test'
+    trap restore_venv_rights EXIT
+
+    if ! [[ -f "$VENV_PATH/bin/activate" ]]; then
+        prepare_venv
+    else
+        source "$VENV_PATH"/bin/activate
+    fi
+
+    find \( -name __pycache__ -o -name '*.pyc' \) -delete
+
+    trap forward_sigterm SIGTERM
+    trap forward_sigterm SIGINT
+
+    "$@" &
+    child="$!"
+    fg 2 >/dev/null
```

Comment:
Hmmm.... if using fg, not sure if the forwarding of the signals is needed (with just wait it's needed for sure), have to investigate.